src/SHA/Keccak.mo

```diff
@@ -0,0 +1,242 @@
+        0x8000000000008080, 0x0000000080000001, 0x8000000080008008
+    ];
+
+    // Keccak constants
+    private let keccakf_rotc : [Nat64] = [
+        1,  3,  6,  10, 15, 21, 28, 36, 45, 55, 2,  14,
+        27, 41, 56, 8,  25, 43, 62, 18, 39, 61, 20, 44
+    ];
+
+    // Keccak constants
+    private let keccakf_piln : [Nat] = [
+        10, 7,  11, 17, 18, 3, 5,  16, 8,  21, 24, 4,
+        15, 23, 19, 13, 12, 2, 20, 14, 22, 9,  6,  1
+    ];
+
+    public class Keccak(
+      initialState   : [Nat64], // unused for now
+      hashSize       : Nat,
+      delimitedSuffix: Nat8,
+    ) : Hash.Hash = {
+        // Keccak l-parameter selects "bus size". For SHA3, l = 6 and
+        // this is the only value supported at this time.
+        private let keccakf_l : Nat = 6;
+
+        // Number of rounds
+        private let n_rounds: Nat = 12 + 2*keccakf_l;
+
+        // Bus sizes in bits
+        private let bsize: Nat = 25 * 2**keccakf_l;
+
+        // Capactiy in bits
+        private let cap: Nat = hashSize * 2;
+
+        // Block size in bytes
+        assert(bsize > cap+8);
+        private let rsize: Nat = (bsize - cap)/8;
+
+        // Internal state
+        private let state: [var Nat64] = Array.init<Nat64>(bsize/64, 0);
+
+        // Current write state index
+        private var pt: Nat = 0;
+
+        // blockSize in bytes
+        public func blockSize() : Nat { rsize; };
+
+        // Function to initialize
+        public func reset() : () {
+            for (i in Iter.range(0, state.size()-1)) {
+                state[i] := 0;
+            };
+            pt := 0;
+        };
+
+        // Return size of hash in bits
+        public func size() : Nat { hashSize; };
+
+        // Turn array of Nat8s into Nat64, LSB first
+        private func pack64(data : [Nat8]) : Nat64 {
+            let dsize = data.size();
+            assert(dsize <= 8 and dsize > 0);
+
+            var q: Nat64 = 0;
+            // Avoid reverse() here to optimize? Maybe unroll?
+            for (i in Iter.range(0, dsize-1:Nat)) {
+                q |= Nat64.fromNat(Nat8.toNat(data[dsize-1:Nat-i]));
+                if (i < (dsize-1:Nat)) {
+                    q <<= 8;
+                };
+            };
+            return q;
+        };
+
+        // Unpack Nat64 into array of Nat8s, LSB first
+        private func unpack64(q : Nat64) : [Nat8] {
+            var t = q;
+            let qbuf = Array.init<Nat8>(8, 0);
+            for (i in Iter.range(0, qbuf.size()-1)) {
+                qbuf[i] := Nat8.fromNat(Nat64.toNat(t & 0xff));
+                t >>= 8;
+            };
+            return Array.freeze<Nat8>(qbuf);
+        };
+
+        private func dump() {
+            Debug.print("State");
+            for (i in Iter.range(0, state.size()-1)) {
+                let b = Array.reverse<Nat8>(unpack64(state[i]));
+                Debug.print(Hex.encode(b));
+            };
+        };
```

Comment:
Can be removed.

Reply:
Thought I'd leave it in can more testing (and debugging) need to be done. Easier to have and asked to remove IMHO. Can remove it no problem.